org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/colors/participants/XMLDocumentColorParticipant.java

```diff
@@ -106,25 +91,34 @@ private void doDocumentColor(DOMNode node, List<XMLColorExpression> expressions,
 	}
 
 	@Override
-	public void doColorPresentations(DOMDocument xmlDocument, ColorPresentationParams params,
-			List<ColorPresentation> presentations, CancelChecker cancelChecker) {
+	public void doColorPresentations(DOMDocument xmlDocument, ColorPresentationParams params, List<ColorPresentation> presentations, CancelChecker cancelChecker) {
 		Color color = params.getColor();
-		Range replace = params.getRange();
-		// RGB color presentation
-		presentations.add(toRGB(color, replace));
-		// Hexa color presentation
-		presentations.add(toHexa(color, replace));
+		try {
+			int startOffset = xmlDocument.offsetAt(params.getRange().getStart());
+			var attribute = xmlDocument.findAttrAt(startOffset);
+			Range editRange = XMLPositionUtility.selectAttributeValue(attribute, true);
+			String rangeText = attribute.getValue();
+			var label = rangeText.startsWith("rgb") ? getRGB(color)
+					: getHex(color, rangeText.startsWith("#"));
+			var edit = new TextEdit(editRange, label) {

```

Comment:
I don't quite understand what you're trying to do here, but I think that you can just use `editRange`. eg.

```java
var edit = new TextEdit(editRange, label);
```

After making this change locally, I wasn't able to reproduce the bug with the wrong replace range for the first few seconds.

Reply:
honestly, at one point I ended up adding a bunch of "let's see what happens" shots trying to fix that range bug, because I just couldn't wrap my head around the apparent asynchronicity of how this method runs..? At least it seemed as if upon applying the created `TextEdit`, the range originally assigned had been shifted alongside the added `#`, but I haven't figured out how to run some validation at the actual insertion time.
Long story short, this block didn't have much intent behind it. I'm afraid I'll need a closer look though to understand _why_ without it, the remaining implementation is now playing ball.